### PR TITLE
SEDMv2 triggering interface

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR enables the two pathways that would need to exist here: (i) a local SkyPortal instance that the request is being made on and (ii) an external SkyPortal instance that this instance would take to (like we would for any other system). So we support both pathways.

### CODE REVIEW

File: skyportal/facility_apis/sedmv2.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        "exposure_time",
+        "exposure_counts",
+        "maximum_airmass",
+        "minimum_lunar_distance",
+        "priority",
+        "start_date",
+        "end_date",
+    ]:
+        if param not in request.payload:
+            raise ValueError(f'Parameter {param} required.')
```
Comment (profjsb) [inline skyportal/facility_apis/sedmv2.py:35]: Probably would benefit from some bounds checking here. I suppose it's the job of the receiver of this to deal bounds but if we can ensure that we are only sending over valid requests then it's worth doing the checking here too.

File: skyportal/facility_apis/sedmv2.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to submit.
+        """
+
+        from ..models import FacilityTransaction, DBSession
+
+        validate_request_to_sedmv2(request)
```
Comment (profjsb) [inline skyportal/facility_apis/sedmv2.py:57]: If this raises an error (ie. ill-formed request) then the error will not handled properly. Need to wrap this in a try/except statement and `return self.error(...)` something sensible if an error is raised.

File: skyportal/tests/flaky/test_followup_requests.py [hunk 3]
```diff
@@ -165,6 +165,72 @@ def add_allocation_ztf(instrument_id, group_id, token):
...
+            "instrument_id": instrument_id,
+            "hours_allocated": 100,
+            "pi": "Ed Hubble",
+            '_altdata': '{"access_token": "testtoken"}',
+        },
+        token=token,
+    )
+    assert status == 200
+    assert data["status"] == "success"
+    return data["data"]
```
Comment (profjsb) [inline skyportal/tests/flaky/test_followup_requests.py:183]: why is this test returning something?

File: skyportal/facility_apis/sedmv2.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to delete from the queue and the SkyPortal database.
+        """
+
+        from ..models import DBSession, FollowupRequest, FacilityTransaction
+
+        validate_request_to_sedmv2(request)
```
Comment (profjsb) [inline skyportal/facility_apis/sedmv2.py:112]: same here with error catching (see above)

File: skyportal/facility_apis/sedmv2.py [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+            },
+            "maximum_airmass": {
+                "title": "Maximum Airmass (1-3)",
+                "type": "number",
+                "default": 2.0,
+                "minimum": 1,
+                "maximum": 3,
+            },
+            "minimum_lunar_distance": {
+                "title": "Maximum Seeing [arcsec] (0-180)",
```
Comment (profjsb) [inline skyportal/facility_apis/sedmv2.py:187]: This key and title seem to be about two different properties.

File: skyportal/facility_apis/sedmv2.py [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+                "type": "number",
+                "default": 30.0,
+                "minimum": 0,
+                "maximum": 180,
+            },
+            "priority": {
+                "type": "number",
+                "default": 1.0,
+                "minimum": 1,
+                "maximum": 5,
```
Comment (profjsb) [inline skyportal/facility_apis/sedmv2.py:197]:
```suggestion
                "minimum": 1.0,
                "maximum": 5.0,
```

**Review by profjsb — CHANGES_REQUESTED** [review]
This looks pretty straightforward to me ... some small changes need IMO.

File: skyportal/facility_apis/sedmv2.py (hunk 5 above)
Comment (mcoughlin) [reply skyportal/facility_apis/sedmv2.py:187]: Good catch.

File: skyportal/facility_apis/sedmv2.py (hunk 2 above)
Comment (mcoughlin) [reply skyportal/facility_apis/sedmv2.py:57]: All of the other API's just raise Errors. What class is taking care of the error handling here?

File: skyportal/facility_apis/sedmv2.py (hunk 1 above)
Comment (mcoughlin) [reply skyportal/facility_apis/sedmv2.py:35]: Added!

File: skyportal/tests/flaky/test_followup_requests.py (hunk 3 above)
Comment (mcoughlin) [reply skyportal/tests/flaky/test_followup_requests.py]: Great point! Removed!

File: skyportal/facility_apis/sedmv2.py (hunk 2 above)
Comment (profjsb) [reply skyportal/facility_apis/sedmv2.py:75]: `validate_request_to_sedmv2` either raises an error (`ValueError`) or returns None. Raising this error within an API request will not necessarily return back the sender a response they can parse. [`baselayer` overwrites the tornado `error` method](https://github.com/cesium-ml/baselayer/blob/ec952bbb3fcde91c19a9c2397c38f16e3d6b29c7/app/handlers/base.py#L286) to handle this gracefully.  So I'd do something like:

```
        try:
            validate_request_to_sedmv2(request)
        except (ValueError, ) as e:
            return self.error('SEMDv2 validation error: {e.normalized_messages()}')
```

Comment (mcoughlin) [reply skyportal/facility_apis/sedmv2.py:75]: ok but that self isn't available anywhere here is it? this is a static method called in handlers/api/followup_request.py. The exception ends up in the red bar text on the front end.

Comment (profjsb) [reply skyportal/facility_apis/sedmv2.py:75]: Ah, you're right. I thought that these APIs were subclasses of the API handlers we have in the rest of the app. But instead it's a base of `FollowUpAPI`. So what `self.` is indeed available here `self.error` has not been implemented. I'd argue we should implement it in the `FollowUpAPI` (or its base class) but that's beyond the scope of this PR.

**Review by profjsb — APPROVED** [review]

File: skyportal/facility_apis/sedmv2.py [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+        raise ValueError('exposure_counts must be at least 1.')
+
+    if request.payload["maximum_airmass"] < 1:
+        raise ValueError('maximum_airmass must be at least 1.')
+
+    if (
+        request.payload["minimum_lunar_distance"] < 0
+        or request.payload["minimum_lunar_distance"] > 180
+    ):
+        raise ValueError('maximum_airmass must be within 0-180.')
```
Comment (guynir42) [inline skyportal/facility_apis/sedmv2.py:54]:
```suggestion
        raise ValueError('minimum lunar distance must be within 0-180.')
```

File: skyportal/facility_apis/sedmv2.py [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+
+        Parameters
+        ----------
+        request: skyportal.models.FollowupRequest
+            The request to delete from the queue and the SkyPortal database.
+        """
+
+        from ..models import DBSession, FollowupRequest, FacilityTransaction
+
+        validate_request_to_sedmv2(request)
```
Comment (guynir42) [inline skyportal/facility_apis/sedmv2.py:130]: Do we need to submit all the required parameters even though we are just deleting a followup?

**Review by guynir42 — COMMENTED** [review]
Some small comments.

File: skyportal/facility_apis/sedmv2.py (hunk 8 above)
Comment (mcoughlin) [reply skyportal/facility_apis/sedmv2.py]: Great point. Removed.

**Comment by guynir42** [comment]
Looks good. Rerunning tests just to make sure the one failure was unrelated.

**Review by guynir42 — APPROVED** [review]

**Pull request merged**